### PR TITLE
[JENKINS-57970] Available updates

### PR DESCRIPTION
**Pull request by stopalopa**
Finally added the option for users to see the plugins that will be downloaded before they download them.  Using the --list option, users will be able to see the installed/bundled plugins, the list of all requested plugins and their dependencies, and what plugins will actually be installed by comparing the list of requested plugins against the currently installed plugins. Users can also user the --verbose option to optionally see the details of the plugin information (https://issues.jenkins-ci.org/browse/JENKINS-58225).  

Will be adding support for https://issues.jenkins-ci.org/browse/JENKINS-58191 next. It was requested that the available updates be determined for the *effective plugin set*, which is now being stored explicitly.

Also adds support for https://issues.jenkins-ci.org/browse/JENKINS-57970

### CODE REVIEW

File: plugin-management-cli/src/main/java/io/jenkins/tools/pluginmanager/cli/CliOptions.java [hunk 1]
```diff
@@ -41,8 +41,20 @@
     @Option(name = "--war", aliases = {"-w"}, usage = "Path to Jenkins war file")
     private String jenkinsWarFile;
 
+    @Option(name = "--list", aliases = {"-l"}, usage = "Lists all plugins currently installed and if given a list of " +
+            "plugins to install either via file or CLI option, all plugins that will be installed by the tool",
+            handler = BooleanOptionHandler.class)
+    private boolean showPluginsToBeDownloaded;
+
+    @Option(name = "--verbose", usage="Output more plugin dependency information", handler = BooleanOptionHandler.class)
+    private boolean isVerbose;
```
Comment (timja) [inline plugin-management-cli/src/main/java/io/jenkins/tools/pluginmanager/cli/CliOptions.java]: isBoolean should be on a getter not the field
```suggestion
    private boolean verbose;
```

File: plugin-management-cli/src/main/java/io/jenkins/tools/pluginmanager/cli/CliOptions.java [hunk 2]
```diff
@@ -41,8 +41,20 @@
...
+
+    @Option(name = "--available-updates", aliases = {"-u"}, usage = "Set to true to see if any updates are available for the " +
```
Comment (timja) [inline plugin-management-cli/src/main/java/io/jenkins/tools/pluginmanager/cli/CliOptions.java]:
```suggestion
    @Option(name = "--available-updates", usage = "Show available plugin updates for the " +
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Config.java [hunk 3]
```diff
@@ -30,6 +33,9 @@ private Config(
             File pluginDir,
             boolean showWarnings,
             boolean showAllWarnings,
+            boolean showAvailableUpdates,
+            boolean showPluginsToBeDownloaded,
+            boolean isVerbose,
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Config.java]:
```suggestion
            boolean verbose,
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Config.java [hunk 4]
```diff
@@ -58,6 +67,16 @@ public boolean isShowAllWarnings() {
         return showAllWarnings;
     }
 
+    public boolean isShowAvailableUpdates() {
+        return showAvailableUpdates;
+    }
+
+    public boolean isShowPluginsToBeDownloaded() { return showPluginsToBeDownloaded; }
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Config.java]:
```suggestion
    public boolean isShowPluginsToBeDownloaded() { 
        return showPluginsToBeDownloaded; 
    }
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Config.java [hunk 5]
```diff
@@ -86,9 +105,11 @@ public static Builder builder() {
         private File pluginDir;
         private boolean showWarnings;
         private boolean showAllWarnings;
+        private boolean showAvailableUpdates;
+        private boolean showPluginsToBeDownloaded;
+        private boolean isVerbose;
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Config.java]:
```suggestion
        private boolean verbose;
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 6]
```diff
@@ -141,13 +213,28 @@ public void getSecurityWarnings() {
...
 
+
+    public boolean warningExists(Plugin plugin) {
+        String pluginName = plugin.getName();
+        if (allSecurityWarnings.containsKey(pluginName)) {
+            for (SecurityWarning.SecurityVersion effectedVersion : allSecurityWarnings.get(pluginName)
+                    .getSecurityVersions()) {
+                Matcher m = effectedVersion.getPattern().matcher(plugin.getVersion().toString());
+                if (m.matches()) {
+                    //System.out.println("Security warning for " + plugin);
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: do you want to log this? shouldn't leave commented out code in

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 7]
```diff
@@ -195,14 +282,50 @@ public void outputFailedPlugins() {
...
                 failedPlugins.add(plugin);
             }
         }
     }
 
+    public Map<String, Plugin> findPluginsAndDependencies(List<Plugin> requestedPlugins) {
+        Map<String, Plugin> allPluginDependencies = new HashMap<>();
+
+        for (Plugin requestedPlugin : requestedPlugins) {
+            //for each requested plugin, find all the dependent plugins that be downloaded (including requested plugin)
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
            //for each requested plugin, find all the dependent plugins that will be downloaded (including requested plugin)
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 8]
```diff
@@ -195,14 +282,50 @@ public void outputFailedPlugins() {
...
+                String dependencyName = dependentPlugin.getName();
+                VersionNumber dependencyVersion = dependentPlugin.getVersion();
+                if (!allPluginDependencies.containsKey(dependencyName)) {
+                    allPluginDependencies.put(dependencyName, dependentPlugin);
+                } else {
+                    Plugin existingDependency = allPluginDependencies.get(dependencyName);
+                    if (existingDependency.getVersion().compareTo(dependencyVersion) < 0) {
+                            if (isVerbose) {
+                                System.out
+                                        .println("Version of " + dependencyName + " (" + existingDependency.getVersion()
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: use String.format please, this isn't very readable

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 9]
```diff
@@ -265,42 +388,57 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
...
+            downloadPlugin(plugin, tempFile);
+            String dependencyString = getAttributefromManifest(tempFile, "Plugin-Dependencies");
+            String[] dependencies = dependencyString.split(",");
 
-        for (String dependency : dependencies) {
-            if (dependency.contains("resolution:=optional")) {
-                dependency = dependency.split(";")[0];
-                isPluginOptional = true;
+            if (isVerbose) {
+                System.out.println("/n" + plugin.getName() + " depends on: ");
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
                System.out.println("\n" + plugin.getName() + " depends on: ");
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 10]
```diff
@@ -265,42 +388,57 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
...
-        String dependencyString = getAttributefromManifest(plugin.getFile(), "Plugin-Dependencies");
-        String[] dependencies = dependencyString.split(",");
+    public List<Plugin> resolveDependenciesFromManifest(Plugin plugin) {
         List<Plugin> dependentPlugins = new ArrayList<>();
 
-        System.out.println(plugin.getName() + " depends on: ");
-
-        boolean isPluginOptional = false;
+        try {
+            File tempFile = Files.createTempFile(plugin.getName(), ".jpi").toFile();
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Does this need to be deleted somewhere?

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 11]
```diff
@@ -330,48 +471,54 @@ public void resolveDependencies(Plugin plugin) {
...
+                String dependencyName = p.getName();
+                if (!recursiveDependencies.containsKey(dependencyName)) {
+                    recursiveDependencies.put(dependencyName, p);
+                    queue.add(p);
                 } else {
-                    System.out.println("Skipping already installed dependency " + dependencyName);
+                    Plugin existingDependency = recursiveDependencies.get(dependencyName);
+                    if (existingDependency.getVersion().compareTo(p.getVersion()) < 0) {
+                        if (isVerbose) {
+                        System.out.println("Version of " + dependencyName + " ( " + existingDependency.getVersion()
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: String.format please

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 12]
```diff
@@ -330,48 +471,54 @@ public void resolveDependencies(Plugin plugin) {
...
+
+        recursiveDependencies.put(plugin.getName(), plugin);
+
+        while (queue.size() != 0) {
+            Plugin dependency = queue.poll();
+
+            if (dependency.getDependencies().isEmpty()) {
+                dependency.setDependencies(resolveDirectDependencies(dependency));
             }
+            Iterator<Plugin> i = dependency.getDependencies().iterator();
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: I think a for loop or a forEach would be simpler than an iterator + while?

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 13]
```diff
@@ -381,40 +528,40 @@ public void downloadDependencies(List<Plugin> dependentPlugins) {
...
-    /**
-     * Determines the plugin download url. If a url is specified from the CLI or plugins file, that url will be used
-     * and the plugin verison and Jenkins version will be ignored. If no url is specified, the url will be
-     * determined from the Jenkins update center and plugin name.
-     *
-     * @param plugin to download
-     * @return url to download plugin from
-     */
+
+        /**
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: indenting looks off here

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java (hunk 10 above)
Comment (stopalopa) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Yes, fixed in later commit

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 14]
```diff
@@ -95,8 +101,113 @@ public void start() {
...
+    }
+
+    /**
+     * Finds the final set of plugins that have been
+     * @param pluginsToBeDownloaded
+     * @return
+     */
+    public Map<String, Plugin> findEffectivePlugins(List<Plugin> pluginsToBeDownloaded) {
+        Map<String, Plugin> effectivePlugins = new HashMap<>();
+        for (Plugin plugin : pluginsToBeDownloaded) {
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: You could probably use streams here as well like
`
List<String> copy = list.stream()
  .collect(Collectors.toList());
`
and then keep passing the stream on to add the other plugins.

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 15]
```diff
@@ -141,13 +252,69 @@ public void getSecurityWarnings() {
...
+                    System.out.println(String.format("%s (%s): %s %s", pluginName, plugin.getVersion(),
+                            allSecurityWarnings.get(pluginName).getName(),
+                            allSecurityWarnings.get(pluginName).getMessage()));
+                }
+            }
+        }
+    }
+
+    /**
+     *
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Javadoc? :)

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 16]
```diff
@@ -195,14 +361,55 @@ public void outputFailedPlugins() {
...
+     * @param requestedPlugins
+     * @return
+     */
+    public Map<String, Plugin> findPluginsAndDependencies(List<Plugin> requestedPlugins) {
+        Map<String, Plugin> allPluginDependencies = new HashMap<>();
+
+        for (Plugin requestedPlugin : requestedPlugins) {
+            //for each requested plugin, find all the dependent plugins that will be downloaded (including requested plugin)
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: +1 just as an aside, feel free to markup what the methods are doing.  It'll be very helpful for collaborators in the repo if you add notes in the code.  Think about what could help someone else who's coming into the codebase with fresh eyes; what would help them figure out what's going on quickly.

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 17]
```diff
@@ -195,14 +361,55 @@ public void outputFailedPlugins() {
...
+
+            for (Plugin dependentPlugin : dependencies.values()) {
+                String dependencyName = dependentPlugin.getName();
+                VersionNumber dependencyVersion = dependentPlugin.getVersion();
+                if (!allPluginDependencies.containsKey(dependencyName)) {
+                    allPluginDependencies.put(dependencyName, dependentPlugin);
+                } else {
+                    Plugin existingDependency = allPluginDependencies.get(dependencyName);
+                    if (existingDependency.getVersion().compareTo(dependencyVersion) < 0) {
+                        if (isVerbose) {
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
                        if (verbose) {
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 18]
```diff
@@ -265,29 +472,46 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
...
+                return dependentPlugins;
+            }
+            String dependencyString = getAttributefromManifest(tempFile, "Plugin-Dependencies");
+            String[] dependencies = dependencyString.split(",");
 
-        for (String dependency : dependencies) {
-            if (dependency.contains("resolution:=optional")) {
-                dependency = dependency.split(";")[0];
-                isPluginOptional = true;
+            if (isVerbose) {
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
            if (verbose) {
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 19]
```diff
@@ -265,29 +472,46 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
...
+                System.out.println("\n" + plugin.getName() + " depends on: ");
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Maybe incomplete logging statement?

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 20]
```diff
@@ -265,29 +472,46 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
...
+                dependentPlugins.add(dependentPlugin);
+
+                if (isVerbose) {
+                    System.out.println(pluginName + ": " + pluginVersion);
+                }
+            }
+            Files.delete(tempFile.toPath());
+            return dependentPlugins;
+        } catch (IOException e) {
+            System.out.println("Unable to resolve dependencies");
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: You might want to log this error, and minimally what plugin you were processing when the error occurred.

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 21]
```diff
@@ -296,31 +520,35 @@ public void resolveDependenciesFromManifest(Plugin plugin) {
...
         if (dependencies == null) {
             resolveDependenciesFromManifest(plugin);
-            return;
+            return dependentPlugins;
         }
 
         if (dependencies.length() == 0) {
-            System.out.println(plugin.getName() + " has no dependencies");
-            return;
+            if (isVerbose) {
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
            if (verbose) {
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 22]
```diff
@@ -296,31 +520,35 @@ public void resolveDependenciesFromManifest(Plugin plugin) {
...
+                System.out.println("\n" + plugin.getName() + " has no dependencies");
+            }
+            return dependentPlugins;
         }
 
-        List<Plugin> dependentPlugins = new ArrayList<>();
 
-        System.out.println(plugin.getName() + " depends on: ");
+        if (isVerbose) {
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
        if (verbose) {
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 23]
```diff
@@ -330,48 +558,54 @@ public void resolveDependencies(Plugin plugin) {
             Plugin dependentPlugin = new Plugin(pluginName, pluginVersion, isPluginOptional);
             dependentPlugins.add(dependentPlugin);
 
-            System.out.println(pluginName + ": " + pluginVersion);
+            if (isVerbose) {
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
            if (verbose) {
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java (hunk 20 above)
Comment (stopalopa) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: If there are any errors, I am planning on dealing with these by keeping track of the initial plugins in the plugin directory and then restoring those and deleting anything that was downloaded. I haven't gotten to it yet though, so Idk if we should try to include those changes in the upcoming alpha or not and maybe I should just implement something else in the meantime.

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/Plugin.java [hunk 24]
```diff
@@ -22,7 +22,7 @@ public Plugin(String name, String version, String url) {
         this.version = new VersionNumber(version);
         this.url = url;
         this.dependencies = new ArrayList<>();
-        this.dependents = new ArrayList<>();
+        this.parent = this;
```
[the remark this answers is not in the document]
Comment (stopalopa) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/Plugin.java]: That's a good point. It shouldn't at the moment....more or less the only thing I'm really using that for is to print out information about information about if a less high plugin version gets replaced by a higher plugin version.  Also right now, the plugin tree is only getting resolved downward....but maybe I should think about another way to do that.

**Comment by timja** [comment]
(conflicts)

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java (hunk 19 above)
Comment (stopalopa) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: The dependencies are output below. I could just print them all together at the end though and maybe that would be more clear.

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 25]
```diff
@@ -143,11 +144,9 @@ public void start() {
             if (installedVersion == null) {
                 pluginsToDownload.add(plugin);
             } else if (installedVersion.compareTo(plugin.getVersion()) < 0) {
-                if (verbose) {
-                    System.out.println("Installed version (" + installedVersion + ") of " + pluginName +
-                            " is less than minimum required version of " + plugin.getVersion() +
-                            ", bundled plugin will be upgraded");
-                }
+                log("Installed version (" + installedVersion + ") of " + pluginName +
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: would be good to format this too

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 26]
```diff
@@ -497,7 +495,7 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
 
             //not all plugin Manifests contain the Plugin-Dependencies field
             if (StringUtils.isEmpty(dependencyString)) {
-                System.out.println(plugin.getName() + " has no dependencies");
+                log("\n" + plugin.getName() + " has no dependencies");
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Do you need System.LINE_SEPARATOR here? (not sure if java handles \n on different OS sending the right separator?

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 27]
```diff
@@ -905,4 +892,10 @@ public void setInstalledPluginVersions(Map<String, VersionNumber> installedPlugi
     public void setBundledPluginVersions(Map<String, VersionNumber> bundledPlugins) {
         bundledPluginVersions = bundledPlugins;
     }
+
+    public void log(String message) {
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: I think this should be `verbose` or `logVerbose` or `log(message, verbose)`
log isn't what it is doing

**Comment by timja** [comment]
Code looks good but there's lots of conflicts

**Comment by timja** [comment]
(code's not compiling after merge)
```
[2019-07-29T17:29:57.383Z] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project plugin-management-library: Compilation failure: Compilation failure: 

[2019-07-29T17:29:57.383Z] [ERROR] /mnt/agent-workspace/workspace/_installation-manager-tool_PR-44/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:[171,99] incompatible types: <nulltype> cannot be converted to boolean

[2019-07-29T17:29:57.384Z] [ERROR] /mnt/agent-workspace/workspace/_installation-manager-tool_PR-44/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:[178,95] incompatible types: <nulltype> cannot be converted to boolean

[2019-07-29T17:29:57.384Z] [ERROR] /mnt/agent-workspace/workspace/_installation-manager-tool_PR-44/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java:[372,42] method downloadPlugin in class io.jenkins.tools.pluginmanager.impl.PluginManager cannot be applied to given types;

[2019-07-29T17:29:57.384Z] [ERROR]   required: io.jenkins.tools.pluginmanager.impl.Plugin,java.io.File

[2019-07-29T17:29:57.384Z] [ERROR]   found: io.jenkins.tools.pluginmanager.impl.Plugin

[2019-07-29T17:29:57.384Z] [ERROR]   reason: actual and formal argument lists differ in length

[2019-07-29T17:29:57.384Z] [ERROR] -> [Help 1]

```

**Comment by stopalopa** [comment]
I was working on fixing it and got distracted by looking into something else I thought might be a bug, but I'll submit a fix as soon as I'm done.

File: plugin-management-cli/src/main/java/io/jenkins/tools/pluginmanager/cli/CliOptions.java [hunk 28]
```diff
@@ -41,8 +41,21 @@
     @Option(name = "--war", aliases = {"-w"}, usage = "Path to Jenkins war file")
     private String jenkinsWarFile;
 
+    @Option(name = "--list", aliases = {"-l"}, usage = "Lists all plugins currently installed and if given a list of " +
+            "plugins to install either via file or CLI option, all plugins that will be installed by the tool",
+            handler = BooleanOptionHandler.class)
+    private boolean showPluginsToBeDownloaded;
+
+    @Option(name = "--verbose", usage = "Output more plugin dependency information",
```
Comment (timja) [inline plugin-management-cli/src/main/java/io/jenkins/tools/pluginmanager/cli/CliOptions.java]: It's not just about plugin dependency versions is it? How about?
```suggestion
    @Option(name = "--verbose", usage = "Verbose logging",
```

File: plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerTest.java [hunk 29]
```diff
@@ -54,6 +55,213 @@ public void setUpPM() throws IOException {
...
+
+        pm.setPluginInfoJson(pluginVersionJson);
+
+        Plugin browserStackPlugin1 = new Plugin("browserstack-integration", "1.0.0", null, null);
+        JSONArray browserStackPluginJson1 = pm.getPluginDependencyJsonArray(browserStackPlugin1, pluginVersionJson);
+        assertEquals(dependencies1.toString(), browserStackPluginJson1.toString());
+
+        Plugin browserStackPlugin111 = new Plugin("browserstack-integration", "1.1.1", null, null);
+        JSONArray browserStackPluginJson111 = pm.getPluginDependencyJsonArray(browserStackPlugin111, pluginVersionJson);
+        //assertEquals(dependencies111.toString(), browserStackPluginJson111.toString());
```
Comment (kwhetstone) [inline plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerTest.java]: Looks like we're skipping the test?

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 30]
```diff
@@ -514,11 +533,11 @@ public JSONArray getPluginDependencyJsonArray(Plugin plugin, JSONObject ucJson)
                     dependentPlugins.add(dependentPlugin);
                 }
             }
-            logVerbose(dependentPlugins.isEmpty() ?  String.format("%n%s has no dependencies", plugin.getName()) :
+            logVerbose(dependentPlugins.isEmpty() ? String.format("%n%s has no dependencies", plugin.getName()) :
                     String.format("%n%s depends on: %n", plugin.getName()) +
-                        dependentPlugins.stream()
-                                .map(p -> p.getName() + " " + p.getVersion())
-                                .collect(Collectors.joining("\n")));
+                            dependentPlugins.stream()
```
Comment (kwhetstone) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Extra spaces?

**Comment by stopalopa** [comment]
Such a slog. Probably the only thing worse than unit testing is front end. 😝

**Pull request merged**